### PR TITLE
Clean version of wnordmann/sampleapp

### PR DESCRIPTION
**Pull request by theduckylittle**
1. Complete rebase of the wnordmann/sampleapp
   which should have a clean merge atop ultimateredux.
   (List below is wnordmann's work...)

2. Adds Feature table example.
3. Added util function to reproject geojson from one projection
   to another.
4. Tweaks to Sass/CSS handling.

### CODE REVIEW

File: examples/feature-table/app.jsx [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  const addLayerFromGeoJSON = (url, sourceName) => {
+    // Fetch URL
+    fetch(url)
+      .then(
+        response => response.json(),
+        error => console.error('An error occured.', error),
+      )
+      // addFeatures with the features, source name
+      .then(json => store.dispatch(mapActions.addFeatures(sourceName,
+        reprojectGeoJson(json.features, json.crs))));
```
Comment (bartvde) [inline examples/feature-table/app.jsx]: maybe this has been discussed previously, just wondering if we considered doing this instead:
```
store.dispatch(mapActions.addFeatures(sourceName, json.features, json.crs)
```
or maybe simply:
```
store.dispatch(mapActions.addFeatures(sourceName, json)
```
and handling the rest in sdk core?

File: __tests__/root/util.test.js [hunk 2]
```diff
@@ -53,4 +53,44 @@ describe('util', () => {
...
+      },
+    ];
+    const crs3857 = { type: 'name', properties: { name: 'urn:ogc:def:crs:EPSG::3857' } };
+    const crs4326 = { type: 'name', properties: { name: 'urn:ogc:def:crs:EPSG::4326' } };
+    // convert 3857 to 4326
+    expect(util.reprojectGeoJson(features3857, crs3857)).toEqual(feature4326);
+    // pass in 4326 without a projection and get back 4326
+    expect(util.reprojectGeoJson(feature4326)).toEqual(feature4326);
+    // pass in 4326 and destination projection and get back 3857
+    expect(util.reprojectGeoJson(feature4326, crs4326, 'EPSG:3857')).toEqual(features3857);
```
Comment (bartvde) [inline __tests__/root/util.test.js]: I see we can pass in a mixture of GeoJSON CRS and EPSG codes, I'd probably simplify things here and only accept EPSG codes. The app can then simply run with ```json.crs.properties.name``` ?

If not, I'd use readProjection from the format instead of parsing ourselves in the util function.

File: examples/basic/app.jsx [hunk 3]
```diff
@@ -264,6 +269,7 @@ function main() {
       <button className="sdk-btn" onClick={zoomToNullIsland}>Zoom to Null Island</button>
       <button className="sdk-btn" onClick={addRandomPoints}>Add 10 random points</button>
       <button className="sdk-btn blue" onClick={removeRandomPoints}>Remove random points</button>
+      <button className="sdk-btn" onClick={clusterPoints}>Cluster Points</button>
```
Comment (bartvde) [inline examples/basic/app.jsx]: why cluster here as well given we have a dedicated cluster example?

File: examples/examples.css [hunk 4]
```diff
@@ -29,6 +29,11 @@ body {
     height: 480px;
 }
 
+.sdk-map {
+    width: 100%;
+    height: 480px;
+}
+
```
Comment (bartvde) [inline examples/examples.css]: I get a bit confused given we have both the ```.sdk-map``` and ```#map``` selector doing the same

File: examples/feature-table/app.jsx [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+  // This is called by the onClick, keeping the onClick HTML clean
+  const runFetchGeoJSON = () => {
+    const url = './data/airports.json';
+    addLayerFromGeoJSON(url, 'dynamic-source');
+  };
+
+  // Next few functions are all about building the feature Table
+
+  // Next few functions are all about building the feature Table
```
Comment (bartvde) [inline examples/feature-table/app.jsx]: double comment

File: examples/feature-table/app.jsx [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        if (headers.indexOf(temp[j]) < 0) {
+          headers.push(temp[j]);
+        }
+      }
+    }
+    return headers;
+  };
+
+  // Build out the headers based on supplied list of properties
+  const buildTableHeader = (properties) => {
```
Comment (bartvde) [inline examples/feature-table/app.jsx]: ideally this would be a React.Component

File: examples/index.html [hunk 7]
```diff
@@ -35,10 +40,6 @@
         Demonstrate how to use sprites on a map.
     </li>
     <li>
-        <a href="./paint-change/index.html">Paint Change</a><br>
-        Demo how change to the paint object of the mapbox style
-    </li>
-    <li>
```
Comment (bartvde) [inline examples/index.html]: is this intentional?

File: package.json [hunk 8]
```diff
@@ -37,11 +37,14 @@
     "sass-loader": "^6.0.6",
     "style-loader": "^0.18.2",
     "webpack": "^2.2.1",
-    "webpack-dev-server": "^2.4.5"
+    "webpack-dev-server": "^2.4.5",
+    "deep-freeze": "0.0.1",
```
Comment (bartvde) [inline package.json]: we already have deep-freeze in there? same for jest

File: src/components/map.jsx [hunk 9]
```diff
@@ -122,13 +122,15 @@ function configureMvtSource(glSource) {
...
   // parse the new features,
-  // TODO: This should really check the map for the correct projection.
-  const features = GEOJSON_FORMAT.readFeatures(glSource.data, {
-    featureProjection: mapProjection,
-  });
+
+  let features;
+
+  if (glSource.data.features) {
+    const readFeatureOpt = { featureProjection: mapProjection || 'EPSG:3857' };
```
Comment (bartvde) [inline src/components/map.jsx]: not sure if we have this convention already, but if it's optional, maybe name it ```opt_mapProjection``` instead?

File: src/components/map.jsx [hunk 10]
```diff
@@ -141,12 +143,17 @@ function updateGeojsonSource(olSource, glSource, mapProjection) {
...
   }
+
   // clear the layer WITHOUT dispatching remove events.
   vector_src.clear(true);
-  // bulk load the feature data.
-  vector_src.addFeatures(features);
+
+  if (features) {
+    // bulk load the feature data
+    vector_src.addFeatures(features || null);
```
Comment (bartvde) [inline src/components/map.jsx]: why the ```|| null```? Not needed since you check for features

File: src/reducers/map.js [hunk 11]
```diff
@@ -194,6 +194,11 @@ function removeSource(state, action) {
 /** Creates a new state with the data for a
  *  source changed to the contents of data.
  *
+ *  @param {object} state - redux state
+ *  @param {string} sourceName - name of the souce to be changed
+ *  @param {array} data -  List of features to be added to the source
+ *        If blank determined by openlayers
```
Comment (bartvde) [inline src/reducers/map.js]: I'm not really getting this last sentence

File: src/util.js [hunk 12]
```diff
@@ -90,3 +91,37 @@ export function encodeQueryObject(query) {
...
+  * @param projection - string from geoJSON crs (coordinate referance system)
+  * @param destProj - ['EPSG:4326'] string of target projection for the data
+  * @return array of geoJSON feature in destProj
+ */
+export function reprojectGeoJson(data, projection, destProj = 'EPSG:4326') {
+  const GEOJSON_FORMAT = new GeoJsonFormat();
+
+  let crsName;
+  if (projection && projection.properties && projection.properties.name) {
+    crsName = projection.properties.name;
```
Comment (bartvde) [inline src/util.js]: as said before I'd use readProjection from the format instance instead of we decide to keep this code

File: src/util.js [hunk 13]
```diff
@@ -90,3 +91,37 @@ export function encodeQueryObject(query) {
...
+  const writeFeatureOptions = {
+    featureProjection: destProj,
+    dataProjection: destProj,
+  };
+
+  const new_data = {
+    type: 'FeatureCollection',
+    features: data,
+  };
+  // const features = GEOJSON_FORMAT.readFeatures(new_data, readFeatureOptions);
```
Comment (bartvde) [inline src/util.js]: get rid of commented code

File: webpack.config.js [hunk 14]
```diff
@@ -0,0 +1,74 @@
+const webpack = require('webpack');
+const path = require('path');
```
Comment (bartvde) [inline webpack.config.js]: this file looks like the dev config, but in a production webpack config file name? Why is this needed?

File: examples/index.html (hunk 7 above)
Comment (theduckylittle) [inline examples/index.html]: It was a double-entry... I fixed that.

File: __tests__/root/util.test.js (hunk 2 above)
Comment (wnordmann) [inline __tests__/root/util.test.js]: using readProjectionFromObject() way cleaner, good call

File: examples/feature-table/app.jsx (hunk 1 above)
Comment (wnordmann) [inline examples/feature-table/app.jsx]: I got the map reducer addFeatures much smarter now

File: examples/index.html (hunk 7 above)
Comment (wnordmann) [inline examples/index.html]: I think this is a reordering issue with git, the Paint Change example is in the index.html

File: package.json (hunk 8 above)
Comment (wnordmann) [inline package.json]: Cleaned up

File: src/components/map.jsx (hunk 9 above)
Comment (wnordmann) [inline src/components/map.jsx]: I could use ES6 defaults along with your naming convention

File: src/components/map.jsx (hunk 10 above)
Comment (wnordmann) [inline src/components/map.jsx]: removed

File: src/reducers/map.js (hunk 11 above)
Comment (wnordmann) [inline src/reducers/map.js]: Old comment not removed before

File: src/util.js (hunk 12 above)
Comment (wnordmann) [inline src/util.js]: cleaned up

File: src/util.js (hunk 13 above)
Comment (wnordmann) [inline src/util.js]: ok

File: webpack.config.js (hunk 14 above)
Comment (wnordmann) [inline webpack.config.js]: I don't think we need webpack.config.js I think this is a git mixup
